README.md
# cs247-project2


changes.txt
Project 2: Neural Networks for Indentifying Handwritten Digits

Student1: Mohammed Ibrahim
Student2: Oluwatobi Oni-Orisan


===========PART 2==================
We modified the function that checked the values in the predicted set. If the value of every slot except the correct slot was less than 0.3 and the value in the correct slot was greater than 0.7 we declared that to be a "match". Else, we counted it as an error.

===========PART 3==================

Observations:
    - Learning Rate: 0.002
        - Number of Tests: 5000
        - Total Probable Errors: 5000
        - Test Loss: 0.2632
        - Validation Set Loss: 0.2696
    - Learning Rate: 0.01
        - Number of Tests: 5000
        - Total Probable Errors: 5000
        - Test Loss: 0.09939
        - Validation Set Loss: 0.0996
    - Learning Rate: 0.05
        - Number of Tests: 5000
        - Total probable errors: 4994
        - Test Loss: 0.0868
        - Validation Set Loss: 0.0884
    - Learning Rate: 0.2
        - Number of Tests: 5000
        - Total probable errors: 4929
        - Test Loss: 0.0752
        - Validation Set Loss: 0.0804
    - Learning Rate: 1.0
        - Number of Tests: 5000
        - Total probable errors: 3290
        - Test Loss: 0.0372
        - Validation Set Loss: 0.0498
    - Learning Rate: 5.0
        - Number of Tests  5000
        - Total probable errors  1814 
        - Test Loss: 0.0078
        - Validation Set Loss: 0.0355
    - Learning Rate: 20.0
        - Number of Tests  5000
        - Total probable errors  1505
        - Test Loss: 0.0039
        - Validation Set Loss: 0.0322


===========PART 4(Momentum)==================

Observations :
    - Learning Rate: 0.002
        - Number of Tests: 5000
        - Total Probable Errors: 5000
        - Test Loss: 0.090634853
        - Validation Set Loss: 0.0924245
    - Learning Rate: 0.01
        - Number of Tests: 5000
        - Total Probable Errors: 4952
        - Test Loss: 0.082565799
        - Validation Set Loss: 0.084867187
    - Learning Rate: 0.05
        - Number of Tests: 5000
        - Total Probable Errors: 4209
        - Test Loss: 0.04762511
        - Validation Set Loss: 0.060104921
    - Learning Rate: 0.2
        - Number of Tests: 5000
        - Total Probable Errors: 2295
        - Test Loss: 0.019010777
        - Validation Set Loss: 0.039789379
    - Learning Rate: 1.0
        - Number of Tests: 5000
        - Total Probable Errors: 2019
        - Test Loss: 0.0086130574
        - Validation Set Loss: 0.036413942
    - Learning Rate: 5.0
        - Number of Tests: 5000
        - Total Probable Errors: 2308
        - Test Loss: 0.029293643
        - Validation Set Loss: 0.050930604
    - Learning Rate: 20.0
        - Number of Tests: 5000
        - Total Probable Errors: 4564
        - Test Loss: 0.092157245
        - Validation Set Loss: 0.092216231

Comments: We tweaked the learning rate (tried values greater than 1 and less than 2 and also between 2 and 3), the momentum (tried 8.0, 1.0 and 1.5), and the number of hidden layers (tried 100) and we concluded that the best learning rate we found was 1.0 (with the lowest error rate) 

Fine Tuning: We kept testing our learning rate and we arrived at a final value of 1.4 with 10 hidden units and 0.94 momentum


===========PART 4(Generalization)==================
We added L2 regularization using the best parameters we had (10 hidden layers, learning rate of 1.4 and momentum of 0.94) and it did help. The number of errors, and loss were decresed. We were initially using out own definition of l2 regularization, but when we used the built-in function for regularization (tf.nn.l2_loss()), the errors and loss were decreased even further.

Without regularization, we got overfitting with at 30 hidden layers. There was a large deviation between the loss of the test set and that of the validation set.

Without regularization, we found the best hidden layer to be 10 which is the amount we were given from the start. Upon numerous test values (5,10, 15, 20, 25, 30, 35, 40,.....75) we discovered the best value to be to be 10.

With regularization we found the best hidden layer to be 17. We testes a bunch of values (10, 12, 15, 17, 38, 42, 47, 55) and 17 gave the best result with least amount of deviation between the loss of test and training sets, even when we tweaked the learning rate.

===========PART 4(Generalization)==================

Settings:
   - Hidden Layers: 17
   - Learning Rate: 1.4
Test Data Classification Error: 0.00932


classifyMNIST.py
import numpy as np
import tensorflow as tf
import matplotlib.pyplot as plt
import sys



from tensorflow.examples.tutorials.mnist import input_data
mnist = input_data.read_data_sets('MNIST_data', one_hot=True)

print 'Testing   Set count ', mnist.train.images.shape
print 'Training   Set count ', mnist.test.images.shape
print 'Validation Set count ', mnist.validation.images.shape

#the length of each input is determined by the data file

training_size = 55000
epochs = 10000
howOften = 500
batch_size = 10

# default hyperparameters

hiddenLayerSize = 17
learning_rate = 1.4
momentum = 0.94

print sys.argv

for o in range(1,len(sys.argv),2):
    print o
    arg = sys.argv[o]
    print arg
    if arg in ['-h', '-hiddens']:
        hiddenLayerSize = int(sys.argv[o+1])
    elif arg in ['-l','-learningRate']:
        learning_rate = float(sys.argv[o+1])
    elif arg in ['-e', '-epochs']:
        epochs = int(sys.argv[o+1])
    elif arg in ['-t', '-trainsize']:
        training_size = int(sys.argv[o+1])
    elif arg in ['-b', '-batchsize']:
        batch_size = int(sys.argv[o+1])

points = mnist.train.images
pointsA = mnist.train.labels #Proper points labels


validation = mnist.validation.images
validationA = mnist.validation.labels #Validation Answers

inputLayerSize = len(points[0])
outputLayerSize = 10

print "epochs            ", epochs
print "input layer size  ", inputLayerSize
print "hidden layer size ", hiddenLayerSize
print "learning rate     ", learning_rate
print "training size     ", len(points)
print "validation size   ", len(validation)


def display_digit(X):
    image = X.reshape([28,28])
    plt.imshow(image, cmap=plt.get_cmap('gray_r'))
    plt.show()
    
# some number of inputs, each of which is the size of the input layer

x = tf.placeholder(dtype=tf.float32, shape=[None, inputLayerSize], name="inputData")
y = tf.placeholder(dtype=tf.float32, shape=[None, outputLayerSize], name="outputData")

weightsInHid = tf.Variable(tf.random_normal([inputLayerSize, hiddenLayerSize], dtype=tf.float32), name='weightsInHid')
biasesHid = tf.Variable(tf.zeros([hiddenLayerSize]), name='biasesHid')
HidIn = (tf.matmul(x, weightsInHid) + biasesHid)
encoded = tf.nn.sigmoid(HidIn)

weightsHidOut = tf.Variable(tf.random_normal([hiddenLayerSize, outputLayerSize], dtype=tf.float32), name='weightsHidOut')
biasesOut = tf.Variable(tf.zeros([outputLayerSize]), name='biasesOut')
decoded = tf.nn.sigmoid(tf.matmul(encoded, weightsHidOut) + biasesOut)

loss = (tf.reduce_mean(tf.square(tf.sub(y, decoded))))
lambdaConstant = 0.000001
#l2reg = tf.reduce_sum(tf.square(weightsInHid)) + tf.reduce_sum(tf.square(weightsHidOut))
l2reg = tf.nn.l2_loss(tf.square(weightsInHid)) + tf.nn.l2_loss(tf.square(weightsHidOut))
print "==================================="

loss = loss + (lambdaConstant* l2reg)
train_op = tf.train.MomentumOptimizer(learning_rate,momentum).minimize(loss)



num_samples = len(points)
print 'Number of Samples',num_samples

sum1 = tf.summary.scalar("training loss", loss)
sum2 = tf.summary.scalar("total loss", loss)

init = tf.global_variables_initializer()
sess = tf.Session() 
sess.run(init)

test_writer = tf.summary.FileWriter('./test', sess.graph)


def checkErrors(ins,outs,flag=False):
    errors = 0
    print "Number of Tests ",len(ins)
    for k in range(len(ins)):
        l, d = sess.run([loss,decoded], feed_dict={x: [ins[k]], y:[outs[k]]})
        # Classified as an error if the value of the invalid buckets is less than 0.3 
        # and the value of the valid bucket is greater than 0.7.
        isError = False
        highCutOff = 0.7
        lowCutOff = 0.3
        for i in range(len(d[0])):
            if outs[k][i] == 1:
                if d[0][i] < highCutOff:
                    isError = True
            else:
                if d[0][i] > lowCutOff:
                    isError = True
        if isError:
            errors = errors +1
        if flag:
            print "test number and error", k, l
            print "predicted"
            print d
            print "desired"
            print outs[k]
            
    print "Total probable errors ", errors



for i in range(epochs):
    trainPoints = []
    trainPointsA = []
    for j in range(batch_size):
        r = np.random.randint(0,num_samples)
        trainPoints.append(points[r])
        trainPointsA.append(pointsA[r])
                          
    l, _ = sess.run([loss, train_op], feed_dict={x: trainPoints, y:trainPointsA})
    write1 = sess.run(sum1, feed_dict={x:trainPoints,y:trainPointsA})
    test_writer.add_summary(write1,i*epochs+j)
    if i % howOften == 0:
        print 'epoch ',i
        big_loss = sess.run(loss,feed_dict={x:points,y:pointsA})
        valid_loss = sess.run(loss,feed_dict={x:validation,y:validationA})
        #print 'Total loss', big_loss
        #print 'Validation Set Loss', valid_loss 
        write2 = sess.run(sum2, feed_dict={x:points,y:pointsA})
        test_writer.add_summary(write2,i)

        print 'Test Set Errors'
        checkErrors(points,pointsA)
        print "Validation errors"
        checkErrors(validation, validationA)

print 'Total loss', sess.run(loss,feed_dict={x:points,y:pointsA})
checkErrors(points,pointsA)

print 'Validation Set Loss', sess.run(loss,feed_dict={x:validation,y:validationA}) 
checkErrors(validation, validationA,False)

exit()

